src/icmp.rs

```diff
@@ -0,0 +1,353 @@
+        }
+    }
+}
+
+fn parse_icmp_code(input: &[u8]) -> IResult<&[u8], IcmpCode> {
+    let (input, code) = number::streaming::be_u16(input)?;
+
+    Ok((input, code.into()))
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum IcmpData {
+    Unreachable { nexthop_mtu: u16, header: IPv4Header, packet: [u8; 8] },
+    Redirect {gateway: Ipv4Addr, header: IPv4Header, packet: [u8; 8] },
+    TimeExceeded {header: IPv4Header, packet: [u8; 8] },
+    None,
+}
+
+fn parse_ipv4_header_and_packet(input: &[u8]) -> IResult<&[u8], (IPv4Header, [u8; 8])> {
+    let (input, header) = parse_ipv4_header(input)?;
+    if input.len() < 8 {
+        return Err(Err::Incomplete(Needed::Size(8 - input.len())));
+    }
+
+    let mut packet: [u8; 8] = Default::default();
+    packet.copy_from_slice(&input[..8]);
+    Ok((&input[8..], (header, packet)))
+}
+
+fn parse_icmp_unreachable_data(input: &[u8]) -> IResult<&[u8], IcmpData> {
+    let (input, _) = number::streaming::be_u16(input)?;
+    let (input, nexthop_mtu) = number::streaming::be_u16(input)?;
+    let (input, (header, packet)) = parse_ipv4_header_and_packet(input)?;
+
+    Ok((input, IcmpData::Unreachable{nexthop_mtu, header, packet}))
+}
+
+fn parse_icmp_redirect_data(input: &[u8]) -> IResult<&[u8], IcmpData> {
+    let (input, gateway) = address(input)?;
+    let (input, (header, packet)) = parse_ipv4_header_and_packet(input)?;
+
+    Ok((input, IcmpData::Redirect{gateway, header, packet}))
+}
+
+fn parse_icmp_timeexceeded_data(input: &[u8]) -> IResult<&[u8], IcmpData> {
+    let (input, _) = number::streaming::be_u32(input)?;
+    let (input, (header, packet)) = parse_ipv4_header_and_packet(input)?;
+
+    Ok((input, IcmpData::TimeExceeded{header, packet}))
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct IcmpHeader {
+    pub code: IcmpCode,
+    pub checksum: u16,
+    pub data: IcmpData,
+}
+
+pub fn parse_icmp_header(input: &[u8]) -> IResult<&[u8], IcmpHeader> {
+    let (input, code) = parse_icmp_code(input)?;
+    let (input, checksum) = number::streaming::be_u16(input)?;
+
+    let (input, data) = match code {
+        IcmpCode::DestinationUnreachable(_) => parse_icmp_unreachable_data(input)?,
+        IcmpCode::Redirect(_) => parse_icmp_redirect_data(input)?,
+        IcmpCode::TimeExceeded(_) => parse_icmp_timeexceeded_data(input)?,
+        _ => (input, IcmpData::None),
+    };
+
+    Ok((
+        input,
+        IcmpHeader {
+            code,
+            checksum,
+            data
+        })
+    )
+}
+
+#[cfg(test)]
+mod tests {
+    use super:: {IcmpHeader, IcmpCode, Unreachable, IcmpData, Redirect, parse_icmp_header};
+    use crate::ipv4::IPv4Header;
+    use crate::ip::IPProtocol;
+    use std::net::Ipv4Addr;
+    use nom::{Err, Needed};
+
+    const EMPTY_SLICE: &'static [u8] = &[];
```

Comment:
odd

Reply:
I took it from other tests. 
Maybe let's keep it in this way so all the tests look the same?

Reply:
no strong feeling